### ISSUE TITLE
asSEXP may need 'PreserveObject'

### ISSUE THREAD

**Issue by randy3k** [JuliaStats/RCall.jl#14]
https://github.com/JuliaStats/RCall.jl/blob/b4afdd6d76f3e2f690cace4b242b1d9c0643c767/src/sexp.jl#L140

I believe that we have to somehow preserve the object created by Rf_allocVector.
It should be fine if this R object is blinded to some variable in some environment. But it may not be always the case.
For example, the following produces segfault
```
x  = RCall.asSEXP([1:10^8])
for i in 1:10
   Reval(Rparse("gc()"))
end
Rprint(x)
```

**Comment by dmbates** [comment]
You're right.  I'll make the corresponding changes.

**Comment by dmbates** [comment]
This is getting more complicated than I first imagined.  Right now I define the SEXP type in Julia as immutable and consisting only of a Ptr{Void}.  But you can't put a finalizer on such an object, because it corresponds to a bitstype.  One way out is to declare SEXP as a type, not an immutable type.  There will be some cost in performance but probably not a measurable difference.

**Comment by randy3k** [comment]
I am thinking there should be two versions of `asSEXP`. One does the preserving and releasing whereas the other variant does not. The reason I suggest this is because releasing an R object could take a lot of time when the number of preserved objects is large.

For example, if one runs `x  = RCall.asSEXP([1:10])` 10 thousands times. If the preserving produce is included, than there will be ten thousands of which `x`. Of course, 9999 of them will be collected when GC, but you can imagine that this step could be time consuming.

I don't really know the best way to resolve this issue. Perhaps we need more experts to give advices.

**Comment by dmbates** [comment]
Perhaps create an abstract type `AbstractSEXP` with a concrete immutable type `SEXP` and a concrete type `protectedSEXP`?

For the time being I suggest keeping the current scheme (everything is the mutable type SEXP and is protected) until we see if there are performance problems.  I doubt very much that R_ReleaseObject is going to be the bottleneck but we can check.  I looked and R_ReleaseObject does do a linear search on a cons list.

**Comment by randy3k** [comment]
Actually, the bottleneck can be easily seen by 
```
julia> @time for i in 1:100000 RCall.asSEXP([1:10]) end
elapsed time: 0.44483907 seconds (26656472 bytes allocated, 68.26% gc time)

julia> @time for i in 1:100000 RCall.asSEXP([1:10]) end
elapsed time: 144.245652407 seconds (28945552 bytes allocated, 99.90% gc time)
```

The second run is **extremely** slow since GC is freeing the memory allocated by the first run.

**Comment by dmbates** [comment]
Right now I am working on documentation.  If you want to create a PR, please do.

**Comment by randy3k** [comment]
Perhaps we can add an additional argument `own` in asSEXP.

When own is true, the R object is persevered, otherwise, it is unprotected.

**Comment by randy3k** [comment]
I think it could be closed as the PR #59 is coming.
